### PR TITLE
Handle external labels in Store's labels API

### LINKED ISSUE #3639: Inconsistency of the behaviors of handling labelValues requests and external labels
We are using different ways to handle `LabelValues` requests for each store.

For `PrometheusStore` which is the sidecar, it checks whether the label we want exists in external labels, if so just return the value. 

``` Go
func (p *PrometheusStore) LabelValues(ctx context.Context, r *storepb.LabelValuesRequest) (*storepb.LabelValuesResponse, error) {
	externalLset := p.externalLabelsFn()

	// First check for matching external label which has priority.
	if l := externalLset.Get(r.Label); l != "" {
		return &storepb.LabelValuesResponse{Values: []string{l}}, nil
	}

	vals, err := p.client.LabelValuesInGRPC(ctx, p.base, r.Label, nil, r.Start, r.End)
```

For `BucketStore`, we get the label values from the block's index and attach the external label value if it matches.

``` Go
			// Do it via index reader to have pending reader registered correctly.
			res, err := indexr.block.indexHeaderReader.LabelValues(req.Label)
			if err != nil {
				return errors.Wrap(err, "index header label values")
			}

			// Add the external label value as well.
			if extLabelValue, ok := extLabels[req.Label]; ok {
				res = strutil.MergeSlices(res, []string{extLabelValue})
			}

```

For `TSDBStore`, we just ignore the external labels. https://github.com/thanos-io/thanos/blob/master/pkg/store/tsdb.go#L221-L235.

I think we need to define the correct logic of the external labels handling for Labels API to make the whole thing more clear. As we plan to support matchers https://github.com/thanos-io/thanos/issues/3637 so it would be more complex.

### PR DESCRIPTION
**Pull request by yeya24**
Signed-off-by: Ben Ye <yb532204897@gmail.com>

Fixes #3639

This pr unifies the behavior of handling external labels on labels API for different stores.

* [ ] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

### CODE REVIEW

**Comment by yeya24** [comment]
For TSDB store and Prometheus store, the workflow is:
1. If the request time range is outside the store's time range, then return an empty array.
2. Else, return the labels result + the external labels names or external label values

For Block store, external labels are per-block so the current logic is fine.

For Proxy store like multitsdb and querier, its external labels will not be added to the results like other stores. This is also an inconsistency but I think it is fine. If anyone has other ideas about this, please let me know.

File: test/e2e/query_test.go [hunk 1]
```diff
@@ -289,12 +289,12 @@ func TestQueryLabelNames(t *testing.T) {
...
 	// Outside time range.
 	labelNames(t, ctx, q.HTTPEndpoint(), nil, timestamp.FromTime(now.Add(-24*time.Hour)), timestamp.FromTime(now.Add(-23*time.Hour)), func(res []string) bool {
-		return len(res) == 0
+		return len(res) == 2
 	})
 
 	labelNames(t, ctx, q.HTTPEndpoint(), []storepb.LabelMatcher{{Type: storepb.LabelMatcher_EQ, Name: "__name__", Value: "up"}},
 		timestamp.FromTime(now.Add(-time.Hour)), timestamp.FromTime(now.Add(time.Hour)), func(res []string) bool {
-			// Expected result: [__name__, instance, job, prometheus, replica]
+			// Expected result: [__name__, instance, job, prometheus, receive, replica, tenant_id]
```
Comment (bwplotka) [inline test/e2e/query_test.go:297]: I like how you updated the commented code to make it work if someone uncomments 😄 😄 
Can we delete this or actually uncomment? 🤗

File: pkg/store/tsdb.go [hunk 2]
```diff
@@ -214,13 +220,24 @@ func (s *TSDBStore) LabelNames(ctx context.Context, r *storepb.LabelNamesRequest
...
+	}
+	res = strutil.MergeSlices(res, extLabelNames)
 	return &storepb.LabelNamesResponse{Names: res}, nil
 }
 
 // LabelValues returns all known label values for a given label name.
 func (s *TSDBStore) LabelValues(ctx context.Context, r *storepb.LabelValuesRequest) (
 	*storepb.LabelValuesResponse, error,
 ) {
+	withinTimeRange, err := s.checkRequestTimeRange(r.End)
```
Comment (bwplotka) [inline pkg/store/tsdb.go:236]: This is out of scope for this PR? Since it's small it's ok, let's just add this to PR tittle 🤗

**Review by bwplotka — APPROVED** [review]
LGTM 👍🏽 Thanks!

Can you help by adding changelog items for this? 🤗

File: test/e2e/query_test.go (hunk 1 above)
Comment (yeya24) [reply test/e2e/query_test.go:297]: Is it a problem if someone uncomments this 🤔  The E2E test cannot be compiled so we can find the change.

**Comment by kakkoyun** [comment]
@yeya24 Do you want to add anything to this? Or address any comments? We can merge it otherwise.

**Comment by yeya24** [comment]
Sorry for the late reply. Will get back to it this week.

**Comment by bwplotka** [comment]
Friendly ping (:

**Comment by kakkoyun** [comment]
@yeya24 Shall we revive this one?
